pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -618,25 +638,51 @@ func (r *ReconcileMachinePool) syncMachineSets(
 					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
 				}
 				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByKey := make(map[string]*corev1.Taint)
+				gTaintsByID := make(map[hivev1.TaintIdentifier]*corev1.Taint)
 				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByKey[gTaint.Key] = &ms.Spec.Template.Spec.Taints[gIndex]
+					gTaintsByID[identifierForTaint(&gTaint)] = &ms.Spec.Template.Spec.Taints[gIndex]
 				}
-				// Go through the remote MachineSet's taints, replacing overlaps with the generated MachineSet
-				for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
-					if gTaint, foundTaint := gTaintsByKey[rTaint.Key]; foundTaint {
-						if gTaint.Value != rTaint.Value || gTaint.Effect != rTaint.Effect {
-							rMS.Spec.Template.Spec.Taints[rIndex] = *gTaint
-							objectModified = true
-						}
-						delete(gTaintsByKey, rTaint.Key)
+				// Build the list of taints for the remote MachineSet from scratch
+				taints := make([]corev1.Taint, 0)
+				// Go through the remote MachineSet's taints, adding/replacing matches from the generated MachineSet
+				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
+					tid := identifierForTaint(&rTaint)
+
+					// If we need to "delete" this one, skip it, but mark modified
+					if taintsToDelete.Has(tid) {
+						objectModified = true
+						continue
 					}
+
+					if gTaint, foundTaint := gTaintsByID[tid]; foundTaint && gTaint.Value != rTaint.Value {
```

Comment:
BUG: if the `rMS` has duplicates, we'll replace the first one, but then because we deleted it from `gTaintsByID`, subsequent dups will be added as is.